Migration fails with additional superuser roles [BF-2080]

Mulugruntz opened the pull request:
### Proposed changes in this pull request

This adds a validation check. If the target database is found to be hardened by `aiven_gatekeeper`, then we make sure that the migration would not try to migrate unauthorized superuser roles. In which case, we would issue an error and stop the process.

Also, as PG 9.6 is no longer supported, there is no longer the need to have a specific subset of fixtures for replication tests. This PR removes this (in a separate commit).

For the CI, there are some changes when installing the environment:
- New dependency: `aiven-pg-security`.
- `aiven-extras`: the latest tag is used, because there are no GitHub Releases on this repository.
- `aiven-pg-security`: the tag from the latest release is used. Newer pre-releases are ignored.

[BF-2080]

### Type (put an `x` where ever applicable)
- [x] Bug fix: Link to the issue
- [ ] Feature (Non-breaking change)
- [x] Feature (Breaking change)
- [ ] Documentation Improvement
- [ ] Other

### Checklist
Please put an `x` against the checkboxes. Write a small comment explaining if its `N/A` (not applicable)

- [ ] All the tests are passing after the introduction of new changes.
- [ ] Added tests respective to the part of code I have written.
- [ ] Added proper documentation where ever applicable (in code and README.md).

[BF-2080]: https://aiven.atlassian.net/browse/BF-2080?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Commit:
Migration fails with additional superuser roles [BF-2080]

This adds a validation check. If the target database is found to be hardened by aiven_gatekeeper, then we make sure that the migration would not try to migrate unauthorized superuser roles. In which case, we would issue an error and stop the process.

[BF-2080]

Commit:
Remove fixtures for PG >= 10 [BF-2080]

Only PG 10+ is supported, so there is no longer the need to have a specific subset of fixtures for replication tests.

[BF-2080]

alexole commented on aiven_db_migrate/migrate/pgmigrate.py:
```diff
@@ -963,7 +1012,37 @@ def _check_pg_ext(self):
+        if unauthorized_roles:
+            self.log.error(
+                "Superuser roles %r from source database are not allowed in target database. "
+                "Allowed roles: %r.",
```
nit: I think we should not expose more data in the logs than needed. Throwing an error for unauthorized OK, but I would not log allowed in this case.

Mulugruntz replied:
If it's a security issue, it's not a nitpick.
Who has access to these logs? Probably only people who run this command? Or could also be other people?

alexole replied:
Logs can be collected and aggregated somewhere. Actually when you ask that, I'd just remove mentioning roles from the logs in general, just leave generic message. Error with info will be propagated to the client anyway.

alexole commented on ci/install.sh:
```diff
@@ -28,9 +35,14 @@ do
+    make -C /aiven-pg-security PG_CONFIG="/usr/lib/postgresql/${dest}/bin/pg_config" clean install
+
+    # Count we have 2 entries: aiven_gatekeeper.so and aiven_extras.so.
+    $(ls /usr/lib/postgresql/${dest}/lib/aiven_{gatekeeper,extras}.so | wc -l | grep -q 2) || exit 1
```
nit: why is it so strict that we need to check for 2 lines exactly? nothing serious but will need adjustments in case of symlinks whatnot will be added

kathia-barahona commented on aiven_db_migrate/migrate/pgmigrate.py:
```diff
@@ -963,7 +1012,33 @@ def _check_pg_ext(self):
 
-    def _check_pg_lobs(self):
+    def _check_aiven_pg_security_agent(self) -> None:
+        if not self.target.has_aiven_gatekeeper:
+            return
+
+        self.log.info("Aiven PostgreSQL security agent is enabled in target.")
```
should we check `self.target.is_pg_security_agent_enabled` instead? Otherwise the log message is not correct

kathia-barahona replied:
I still think its a bit misleading :sweat_smile: since you are stating that we might do it when it is disabled. It is still saying it is enabled.  Rather have smth like `Aiven PostgreSQL gatekeeper is enabled in target` instead.

Mulugruntz replied:
Hmmm... Did you see the updated message? It's now `self.log.info("Aiven PostgreSQL security agent is used by the target.")`

We should probably resolve that conversation and potentially open a new one on the new line, to avoid any confusion :)

kathia-barahona replied:
That is what I mean. The condition of the if statement does not match with what it is being logged. I understand what you are saying, but the code does not reflect this. So, in a future it might confuse someone else without a previous context

Mulugruntz replied:
Well, maybe I'm missing your point. But if it's in the shared preload libs, that means it's used. Even though it could be enabled or disabled (settings). I think logging "enabled" when it could be disabled would be super confusing, don't you think so?

kathia-barahona replied:
so, my point is that it is a bit confusing that we verify if the target has Aiven Gatekeeper as a setting, and logging that  the security agent is been used. I understand its because of transitive property, but maybe a comment will help others to understand why we are logging that message.

Mulugruntz replied:
@kathia-barahona 
Okay, fair enough. If you feel like it might be confusing, it's valid.
I added a clean docstring to explain this.
Tell me what you think of it.

Commit:
Move `PGRunner` to `test/utils.py` [BF-2080]

To no longer import from `conftest.py` and avoid circular imports.

[BF-2080]

Commit:
Test teardown drops logical slots before dropping DB [BF-2080]

We do this because in some rare cases, the teardown already drops the table while the replication slot is still active, which causes the drop to fail. In the cleanup, it is executed after all connections are closed, so it should always succeed.

[BF-2080]

Mulugruntz commented:
@alexole @kathia-barahona 
To be noted that the integration PR passed the tests.

kathia-barahona commented on aiven_db_migrate/migrate/pgmigrate.py:
```diff
@@ -107,6 +107,7 @@ class PGCluster:
     _pg_lang: Optional[List[Dict[str, Any]]]
     _pg_roles: Dict[str, PGRole]
     _mangle: bool
+    _has_aiven_gatekeper: Optional[bool]
```
Could you please link an issue here :)

Mulugruntz replied:
https://aiven.atlassian.net/browse/BF-2150

kathia-barahona commented on aiven_db_migrate/migrate/pgmigrate.py:
```diff
@@ -963,7 +1012,33 @@ def _check_pg_ext(self):
+        unauthorized_roles = su_missing_roles - reserved_roles
+
+        if unauthorized_roles:
+            self.log.error("Some superuser roles from source database are not allowed in target database.")
```
Alright, makes sense that we don't want to expose that information

kathia-barahona commented:
lgtm,  but there is a failed check related to unit tests

alexole commented on ci/install.sh:
```diff
@@ -14,12 +14,19 @@ apt-get install -y \
+
+# Clone aiven-pg-security (aiven_gatekeeper): https://github.com/aiven/aiven-pg-security/releases/tag/v1.0.4
+git clone https://github.com/aiven/aiven-pg-security/ /aiven-pg-security
+git -C /aiven-pg-security checkout 3107f450ee765515a8a19c78e4f3125ce6ab1f36
```
This one is obsolete, I'd rather checkout main or latest tag (which is v1.0.6 - Stable), which should be nice to have somewhere explicitly (top of shell script as var?) 
Also probably would be nice to bump aiven-extras as well.

Mulugruntz replied:
Great idea. Have a look at the new commit I pushed.

- `aiven-extras`: the latest tag is used, because there are no GitHub Releases on this repository.
- `aiven-pg-security`: the tag from the latest release is used. Newer pre-releases are ignored.

Commit:
Improve `ci/install.sh` to use latest deps [BF-2080]

The two aiven dependencies are now picked up in their latest versions, instead of commit hash.

- `aiven-extras`: the latest tag is used, because there are no GitHub Releases on this repository.
- `aiven-pg-security`: the tag from the latest release is used. Newer pre-releases are ignored.

[BF-2080]
